File: HW05/Circle.cpp
#include "Circle.h"



namespace MyShape{

	
	double Circle::total_area=0;
	double Circle::total_perimeter=0;

	Circle::Circle(double rad):Shape(),radius(rad){ }

	Circle::Circle():Shape(),radius(1.){  }

	double Circle::area()const throw(){ return PI*radius*radius; }
	double Circle::perimeter()const throw(){ return 2.*PI*radius; }

	double Circle::getRadius()const throw(){ return radius; }
	double Circle::getX() const throw(){ return x; }
	double Circle::getY() const throw(){ return y; }

	void Circle::setRadius(double rad) throw(ExceptionErrorr){ 
		try{
			if(rad<0){
				throw ExceptionErrorr(rad);
			}
			radius=rad; 
		
		}
		catch(ExceptionErrorr a){
			cout<<endl<<"Radius cannot be negative"<<endl;
		} 
	
	}
	void Circle::setX(double _x) throw(ExceptionErrorr){ 

		try{
			if(_x<0){	
				throw ExceptionErrorr(_x);
			}
			x=_x; 
		}
		catch(ExceptionErrorr a){
			cout<<endl<<"x:"<<a.getMessage()<<" cannot be negative";
		}
	
	}
	void Circle::setY(double _y) throw(ExceptionErrorr)
	{
		try{
			if(_y<0)
				throw ExceptionErrorr(_y);
		 y=_y; 
		}
		catch(ExceptionErrorr a){
			cout<<endl<<"x:"<<a.getMessage()<<" cannot be negative";
		}
	}

	Shape& Circle::operator++() throw(){
		x++;
		y++;
		return (*this);
	}


	/*!!!!!!!!!!!!!!!!!WARNING!!!!!!!!!!!!!!!!!
	*
	*	If you use this operator you have 
	*	to delete return adress manually!!!!!!!!!
	*
	*/
	Shape& Circle::operator++(int IgnoreMe) throw(){
		Circle *circle = new Circle(*this);
			x++;
			y++;
		return (*circle);
	}

	Shape& Circle::operator--() throw( ){		
			x--;
			y--;
			return (*this);
	}


	/*!!!!!!!!!!!!!!!!!WARNING!!!!!!!!!!!!!!!!!
	*
	*	If you use this operator you have 
	*	to delete return adress manually!!!!!!!!!
	*
	*/
	Shape& Circle::operator--(int IgnoreMe) throw( ){
		Circle *circle = new Circle(*this);
			x--;
			y--;
			return (*circle);
	}

	/*overloaded << operator*/
	ostream& operator<<(ostream &outputStream,Circle circle){
		outputStream << " <circle cx=\""<< circle.getX() <<"\" cy=\""<< circle.getY() <<"\" r=\""<<circle.getRadius()
					<<"\" fill=\"blue\" /> " << endl;
		return outputStream;
	}

	void Circle::draw(ostream &outputStream) throw(){
		outputStream << " <circle cx=\""<< x <<"\" cy=\""<< y <<"\" r=\""<<radius
					<<"\" fill=\"blue\" /> " << endl;
	}

	/*for static total_area and total_perimeter.*/
	void Circle::increase_total_area(){
		total_area+= PI*radius*radius;
	}
	void Circle::increase_perimeter(){
		total_perimeter+=2*PI*radius;
	}

	/*total area and total perimeter length getter.*/
	double Circle::show_total_area(){
		return total_area;
	}
	double Circle::show_total_perimeter(){
		return total_perimeter;
	}
}

File: HW07/Circle.java
import java.awt.Color;
import java.awt.Graphics;
import java.lang.IllegalArgumentException;


/**
 * 
 * A simple Circle class that 
 * implements shape methods.
 * 
 * @author Yusuf Can Kan
 *
 */
public class Circle implements Shape, Cloneable {
	
	/**
	 * radius of circle
	 */
	private double radius;
	
	/**
	 * X coordinates.
	 */
	private double x;
	
	/**
	 * Y coordinates.
	 */
	private double y;
	
	/**
	 * The area of triangles which created so far.
	 */
	private static double total_area=0;
	
	/**
	 * The perimeter of triangles which created so far.
	 */
	private static double total_perimeter=0;
	
	/**
	 * Simple constructor that sets radius.
	 * @param rad radius of circle.
	 */
	public Circle(double rad){
		radius=rad;
		this.increase_total_area();
		this.increase_total_perimeter();
	}
	
	/**
	 * Simple constructor that sets radius to 1.
	 */
	public Circle(){
		radius=1;
		x=0;
		y=0;
		this.increase_total_area();
		this.increase_total_perimeter();
	}
	
	/**
	 * Getter method for radius.
	 * @return returns radius.
	 */
	public double getRadius(){return radius;}
	
	/**
	 * Getter method for x.
	 * @return x coordinate.
	 */
	public double getX(){return x;}
	
	/**
	 * Getter method for y.
	 * @return y coordinates.
	 */
	public double getY(){return y;}

	/**
	 * Setter method for radius.
	 * @param radius_ radius of rectangle.
	 * @throws IllegalArgumentException throws if radius is smaller than 0.
	 */
	public void setRadius(double radius_)
			throws IllegalArgumentException{
		if(radius_ < 0){
			throw new IllegalArgumentException();
		}
		
		radius = radius_;
	}
	
	/**
	 * Setter method for x coordinate.
	 * @param x_ x coordinate.
	 * @throws IllegalArgumentException throws if coordinate is smaller than 0.
	 */
	public void setX(double x_)
			throws IllegalArgumentException{
		if(x_<0)
			throw new IllegalArgumentException();
		
		x = x_;
	}
	
	/**
	 * Setter method for y coordinate.
	 * @param y_ y coordinate.
	 * @throws IllegalArgumentException throws if coordinate is smaller than 0.
	 */
	public void setY(double y_)
			throws IllegalArgumentException{
		if(y_<0)
			throw new IllegalArgumentException();
		
		y = y_;
	}
	
	@Override
	public double area() {
		return radius*radius*(Math.PI);
	}

	@Override
	public double perimeter() {
		return 2*(Math.PI)*radius;
	}

	@Override
	public void increment() {
		x++;
		y++;
	}

	@Override
	public void decrement() {
		if( (x-1) >= 0) x--;
		if( (y-1) >= 0) y--;
	}

	@Override
	public void draw(Graphics g) {
		g.setColor(Color.ORANGE);
		g.fillOval((int)x, (int)y, (int)radius*2 , (int)radius*2);
		g.setColor(Color.black);
		g.drawOval((int)x, (int)y, (int)radius*2 , (int)radius*2);
	}

	@Override
	public int compareTo(Shape o) {
		if( o.area() == this.area() ) return 1;
		else return 0;
	}
	
	/*
	 * Increases total area.
	 */
	public void increase_total_area(){
		total_area+= radius*radius*(Math.PI);
	}
	
	/*
	 * Increases total perimeter.
	 */
	public void increase_total_perimeter(){
		total_perimeter += 2*(Math.PI)*radius;
	}
	
	/**
	 * returns the total area.
	 * @return total area.
	 */
	public double show_total_area(){
		return total_area;
	}
	
	/**
	 * Returns the total perimeter.
	 * @return total perimeter.
	 */
	public double show_total_perimeter(){
		return total_perimeter;
	}
	
	/**
	 * Clone method.
	 */
	public Circle clone() 
			throws CloneNotSupportedException{ 
		return (Circle)super.clone(); 
	}
	

}
